crates/federate/src/worker.rs

```diff
@@ -0,0 +1,211 @@
+    if id == latest_id {
+      // no more work to be done, wait before rechecking
+      tokio::select! {
+        () = sleep(Duration::from_secs(10)) => { continue; },
+        () = stop.cancelled() => { return Ok(()); }
+      }
+    }
+    let mut processed_activities = 0;
+    'batch: while id < latest_id
+      && processed_activities < SAVE_STATE_EVERY_IT
+      && !stop.is_cancelled()
+    {
+      id += 1;
+      processed_activities += 1;
+      let Some(ele) = get_activity_cached(pool, id).await? else {
+        state.last_successful_id = id;
+        continue;
+      };
+      let (activity, object) = (&ele.0, &ele.1);
+      let inbox_urls = get_inbox_urls(&instance, &site, &followed_communities, activity);
+      if inbox_urls.is_empty() {
+        state.last_successful_id = id;
+        continue;
+      }
+      let Some(actor_apub_id) = &activity.actor_apub_id else {
+        continue; // activity was inserted before persistent queue was activated
+      };
+      let actor = get_actor_cached(pool, activity.actor_type, actor_apub_id).await?;
+
+      let inbox_urls = inbox_urls.into_iter().map(|e| (*e).clone()).collect();
+      let requests = prepare_raw(object, actor.as_ref(), inbox_urls, &data)
+        .await
+        .into_anyhow()?;
+      for task in requests {
+        // usually only one due to shared inbox
+        let mut req = sign_raw(&task, &data, REQWEST_TIMEOUT).await?;
+        tracing::info!("sending out {}", task);
+        while let Err(e) = send_raw(&task, &data, req).await {
+          state.fail_count += 1;
+          state.last_retry = Utc::now();
+          let retry_delay: Duration = retry_sleep_duration(state.fail_count);
+          tracing::info!(
+            "{}: retrying {id} attempt {} with delay {retry_delay:.2?}. ({e})",
+            instance.domain,
+            state.fail_count
+          );
+          stats_sender.send(state.clone())?;
+          FederationQueueState::upsert(pool, &state).await?;
+          req = sign_raw(&task, &data, REQWEST_TIMEOUT).await?; // resign request
+          tokio::select! {
+            () = sleep(retry_delay) => {},
+            () = stop.cancelled() => {
+              // save state to db and exit
+              break 'batch;
+            }
+          }
+        }
+      }
+      // send success!
+      state.last_successful_id = id;
+      state.fail_count = 0;
+    }
+
+    if Utc::now() - last_state_insert
+      > chrono::Duration::from_std(SAVE_STATE_EVERY_TIME).expect("not negative")
+    {
+      last_state_insert = Utc::now();
+      FederationQueueState::upsert(pool, &state).await?;
+      stats_sender.send(state.clone())?;
+    }
+    {
+      // update communities
+      if (Utc::now() - last_incremental_communities_fetch) > chrono::Duration::seconds(10) {
+        // process additions every 10s
+        followed_communities.extend(
+          get_communities(pool, instance.id, &mut last_incremental_communities_fetch).await?,
+        );
```

Comment:
I've changed the delays to be slower because you're righht that it doesn't need to happen every 10s. But the full update is pretty expensive (it fetches 1000+ communities for every instance) compared to the incremental one (usually fetching 0 communities) so I think it's better to keep it?